### ISSUE TITLE
Stage can retain references in __stack

### ISSUE THREAD

**Issue by Leander1P** [openfl/openfl#1274]
If you do `stage.focus = null` somewhere (e.g. in `DisplayObject.__setStageReference( null )` as a result of removing from stage), `Stage.set_focus` will retain a reference to the interactive objects in `__stack`, causing these objects to "leak" until next `stage.focus = somethingElse`.

We're seeing this keep `TextField`s alive, along with all their parents, even if the link to stage is broken high up (e.g. a `DisplayObjectContainer` well up the parent/child tree).

Locally, we have a patch that drops the `Stage.__stack` member variable and uses local `var __stack`, and this fixes the issue.

Does this seem reasonable?  As far as I can tell, the call sequence shouldn't change even with re-entrant calls to `set_focus` (which should possibly be another issue, since the sequence is a bit wonky if you change focus within e.g. `FOCUS_IN` or `FOCUS_OUT`, but that's a much more elaborate change).  Also, this shouldn't increase number of allocs at all (just how soon the last set is available for GC).

(Can submit a PR with this elsewhere over the next week, but figured an issue and general request for comments would be a good start.)

```
Index: Stage.hx
===================================================================
--- Stage.hx	(revision 1153)
+++ Stage.hx	(revision 1154)
@@ -115,7 +115,6 @@
 	private var __mouseY:Float;
 	private var __renderer:AbstractRenderer;
 	private var __rendering:Bool;
-	private var __stack:Array<DisplayObject>;
 	private var __transparent:Bool;
 	private var __wasDirty:Bool;
 	
@@ -176,7 +175,6 @@
 		#end
 		
 		__clearBeforeRender = true;
-		__stack = [];
 		__mouseOutStack = [];
 		
 		stage3Ds = new Vector ();
@@ -1471,7 +1469,7 @@
 			if (oldFocus != null) {
 				
 				var event = new FocusEvent (FocusEvent.FOCUS_OUT, true, false, __focus, false, 0);
-				__stack = [];
+				var __stack = [];
 				oldFocus.__getInteractive (__stack);
 				__stack.reverse ();
 				__fireEvent (event, __stack);
@@ -1481,7 +1479,7 @@
 			if (__focus != null) {
 				
 				var event = new FocusEvent (FocusEvent.FOCUS_IN, true, false, oldFocus, false, 0);
-				__stack = [];
+				var __stack = [];
 				value.__getInteractive (__stack);
 				__stack.reverse ();
 				__fireEvent (event, __stack);
```

**Comment by Leander1P** [comment]
For ref, this was on openfl 4.0.3, but the issue looks like it has not been fixed in `develop`.

Also, `onTouch` already does something like this with a local `var __stack` masking the member.

**Comment by jgranick** [comment]
I believe I saw the following behavior on Flash Player:

```haxe
var textField = new TextField ();
textField.type = INPUT;
textField.addEventListener (Event.CHANGE, function (_) { trace (textField.text); });

var sprite = new Sprite ();
sprite.addChild (textField);
addChild (sprite);

Timer.delay (function () { removeChild (sprite); }, 5000);
```

If you click on the TextField and begin typing, it receives focus and you get text change events. When it is removed from the stage (as a child of another object) it still receives input if you keep typing.

(I have not compiled the above test code, so my apologies if there's a minor error)

If you add or remove the TextField directly, rather than as a child of another object, stage focus changes and it does not receive input events anymore.

Could you check this on your end, and see if this is how it behaves? Our current code should behave similar to Flash with this

**Comment by Leander1P** [comment]
We'll try to spin up a flash build environment and test this.  However, in our specific case, we're seeing a situation where `Stage.focus` *has* already gone to `null` -- for whatever reason -- implying that the textfield decided to defocus -- but the reference still exists in `Stage.__stack` because it was never cleaned up after FOCUS_OUT.  It seems like parroting the "retain focus when no longer attached to stage" compatibility with flash would be a different issue, given that how `Stage.focus` is set and cleared should have nothing to do with `__stack`?  No one ever uses `__stack` without doing `__stack = []` first, which implies that no one needs its contents between calls (the `_fireEvent` is called with a ref, but since everyone else who touches it resets it first, any event-dispatch-in-progress should keep its local stack ref...)

Put simply, once focus is *already* null, should the __stack still hold refs, even if they're apparently never used from there again?

**Comment by Leander1P** [comment]
(A possible alternate fix, to illustrate, would be an "else" clause for the __focus==null case that just did `__stack = []`, which would also fix this "leak".)

**Comment by jgranick** [comment]
Oh, I see now. Yeah, that's unintended.

I think (long ago) the idea was a single `__stack` variable to be re-used. Perhaps we should do that in the future to reduce GC, but for now, we can just make this a temporary variable

**Comment by Leander1P** [comment]
Thanks!

It's hard to get re-use right when in a possibly-reentrant scenario like that (i.e. a listener on FOCUS_OUT or FOCUS_IN can change focus, causing set_focus to be called while set_focus is already on the stack).  As it stands, the `= []` would have been creating a new array and discarding the old one anyway, if I understand things correctly?  So it wouldn't have been reducing GC load anyway.

Seems like EventDispatcher shows that added complexity pretty well, re reentrancy.  (I still need to split that EventDispatcher refactor into a patch series of low-change bugfixes / safety and then slow evolution to the refactored form, on that note.)

**Comment by jgranick** [comment]
`EventDispatcher` would work simply if we cloned before each dispatch, but that would create a lot of garbage. One idea would be a pool of event dispatch arrays to be used temporarily

**Comment by Leander1P** [comment]
Yep; hence the trying not to touch EventDispatcher's high-level array alloc/copy strategy during the refactor in #1105.  The complexity in that case seems to be there for a very good reason, especially considering the relative frequency of event broadcasts.  =)

Presumably `InteractiveObject.__getInteractive` would need a non-`push`ing variant?  I'm not even entirely sure how this would be done, given that `Array` lacks a `reserve()`/`capacity()` style preallocator -- an abstraction that seems to allow it to go crossplatform more easily.  It seems like a harder problem than the known-size-listener-list issue in `EventDispatcher`, since `_getInteractive` is based on the current structure of the parent/child tree, which can change even more frequently / more dynamically than the listener list, but maybe it boils down to a homomorphic problem...

(Should I open a separate issue for this -- e.g. "stack array caching for Stage.set_focus" -- to hold discussion and possibly some test code for commenting on at some point?  Whatever solution *might* also apply to __onTouch stack too.  Or is set_focus/__onTouch a low-frequency enough event that it's not worth chasing at the moment, until it shows up on someone's profiles more prominently?)

**Comment by jgranick** [comment]
The level of frequency by which different code paths are called vary wildly. I think we should pay attention to the profiler and let it instruct where we should optimize, and take the benefit of simplicity (which helps reduce maintenance load and makes it easier to improve minor behavior discrepancies) rather than worrying too much. If the "stack" behavior seems too hot, then yeah, perhaps we come up with something. If not, lets continue to deal with the bigger performance optimizations and continuing to improve the behavior of areas where OpenFL varies from what is expected

**Comment by Leander1P** [comment]
That sounds like a very sane approach.  =)  Can write up a new issue if it shows up as hot at any point in the future.

Thanks!
